Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
     }
 };
 
+class TextureMapperFlattenedLayer final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit TextureMapperFlattenedLayer(const IntRect& rect, double zNear, double zFar)
+        : m_rect(rect)
+        , m_zNear(zNear)
+        , m_zFar(zFar)
+    { }
+
+    inline IntRect layerRect() const
+    {
+        return m_rect;
+    }
+
+    inline bool needsUpdate() const
+    {
+        return !m_textures.size();
+    }
+
+    void update(TextureMapperPaintOptions& options, const std::function<void(TextureMapperPaintOptions&)>& paintFunction)
+    {
+        double prevZNear, prevZFar;
+        options.textureMapper.getDepthRange(prevZNear, prevZFar);
+        options.textureMapper.setDepthRange(m_zNear, m_zFar);
+
+        IntSize maxTextureSize = options.textureMapper.maxTextureSize();
+        for (int x = m_rect.x(); x < m_rect.maxX(); x += maxTextureSize.width()) {
+            for (int y = m_rect.y(); y < m_rect.maxY(); y += maxTextureSize.height()) {
+                IntRect tileRect({ x, y }, maxTextureSize);
+                tileRect.intersect(m_rect);
+
+                auto surface = options.textureMapper.acquireTextureFromPool(tileRect.size(), { BitmapTexture::Flags::SupportsAlpha });
+                {
+                    SetForScope scopedSurface(options.surface, surface.ptr());
+                    SetForScope scopedOffset(options.offset, -toIntSize(tileRect.location()));
+                    SetForScope scopedOpacity(options.opacity, 1);
+
+                    options.textureMapper.bindSurface(options.surface.get());
+                    paintFunction(options);
+
+                    surface = Ref { *options.surface };
```

Comment:
Still this kind of memory managment seems error-prone. Easy to forget to ref.
Also a few lines before where you have to set options.surface to 'surface.ptr()' is fragile. I would suggest to rework this part.

Reply:
Fixed this. acquireTextureFromPool returns Ref<BitmapTexture> but we can assign it to RefPtr locally and then we can use options.surface without explicit Ref